src/codecs/openexr.rs

```diff
@@ -0,0 +1,489 @@
+        let blocks_in_header = self.selected_exr_header().chunk_count as u64;
+        let channel_count = self.color_type().channel_count() as usize;
+
+        let display_window = self.selected_exr_header().shared_attributes.display_window;
+        let data_window_offset = self.selected_exr_header().own_attributes.layer_position - display_window.position;
+
+        let result = read()
+            .no_deep_data().largest_resolution_level()
+            .rgba_channels(
+                move |_size, _channels| {
+                    // TODO debug_assert_eq!(self.header_index, header_index);
+                    vec![0_f32; display_window.size.area() * channel_count]
+                },
+
+                move |buffer, index_in_data_window, (r,g,b,a_or_1): (f32,f32,f32,f32)| {
+                    let index_in_display_window = index_in_data_window.to_i32() + data_window_offset;
+
+                    // only keep pixels inside the data window
+                    // TODO filter chunks based on this
+                    if index_in_display_window.x() >= 0 && index_in_display_window.y() >= 0
+                        && index_in_display_window.x() < display_window.size.width() as i32
+                        && index_in_display_window.y() < display_window.size.height() as i32
+                    {
+                        let index_in_display_window = index_in_display_window.to_usize("index bug").unwrap();
+                        let first_f32_index = index_in_display_window.flat_index_for_size(display_window.size);
+
+                        buffer[first_f32_index * channel_count .. (first_f32_index + 1) * channel_count]
+                            .copy_from_slice(&[r, g, b, a_or_1][0 .. channel_count]);
+
+                        // TODO white point chromaticities + srgb/linear conversion?
+                    }
+                }
+            )
+            .first_valid_layer() // TODO select exact layer by self.header_index?
+            .all_attributes()
+            .on_progress(|progress| {
+                progress_callback(Progress::new(
+                    (progress*blocks_in_header as f64) as u64, blocks_in_header) // TODO precision errors?
+                );
+            })
+            .from_chunks(self.exr_reader).map_err(to_image_err)?;
+
+        // TODO this copy is strictly not necessary, but the exr api is a little too simple for reading into a borrowed target slice
+
+        // this cast is safe and works with any alignment, as bytes are copied, and not f32 values.
+        unaligned_bytes.copy_from_slice(bytemuck::cast_slice(result.layer_data.channel_data.pixels.as_slice()));
+        Ok(())
+    }
+}
+
+
+
+
+/// Write a raw byte buffer of pixels,
+/// panicking if the buffer is not aligned to `f32`
+/// or if it has an invalid length.
+///
+/// Assumes the writer is buffered. In most cases,
+/// you should wrap your writer in a `BufWriter` for best performance.
+pub fn write_buffer(
```

Comment:
Could you either make this private or move the contents into `ExrEncoder::write_image`?